src/main/cpp/src/hyper_log_log_plus_plus.cu

```diff
@@ -0,0 +1,958 @@
+    auto registers_thread_cache =
+      rmm::device_uvector<int32_t>(num_threads_partial_kernel * num_registers_per_sketch,
+                                   stream,
+                                   cudf::get_current_device_resource_ref());
+    auto group_labels_thread_cache = rmm::device_uvector<int32_t>(
+      num_threads_partial_kernel, stream, cudf::get_current_device_resource_ref());
+
+    {  // add this block to release `hash_col`
+      // 1. compute all the hashs
+      auto input_table_view = cudf::table_view{{input}};
+      auto hash_col         = xxhash64(input_table_view, SEED, stream, mr);
+      hash_col->set_null_mask(cudf::detail::copy_bitmask(input, stream, mr), input.null_count());
+      auto d_hashs = cudf::column_device_view::create(hash_col->view(), stream);
+
+      // 2. execute partial group by
+      int64_t num_blocks_p1 =
+        cudf::util::div_rounding_up_safe(num_threads_partial_kernel, block_size);
+      partial_group_sketches_from_hashs_kernel<num_hashs_per_thread>
+        <<<num_blocks_p1, block_size, 0, stream.value()>>>(*d_hashs,
+                                                           group_labels,
+                                                           precision,
+                                                           sketches_output.begin(),
+                                                           registers_thread_cache.begin(),
+                                                           group_labels_thread_cache.begin());
+    }
+    // 3. merge the intermidate result
+    auto num_merge_threads = num_registers_per_sketch;
+    auto num_merge_blocks  = cudf::util::div_rounding_up_safe(num_merge_threads, block_size);
+    merge_sketches_vertically<block_size>
+      <<<num_merge_blocks, block_size, block_size, stream.value()>>>(
+        num_threads_partial_kernel,  // num_sketches
+        num_registers_per_sketch,
+        sketches_output.begin(),
+        registers_thread_cache.begin(),
+        group_labels_thread_cache.begin());
+  }
+
+  // 4. create output columns
+  auto num_long_cols      = num_registers_per_sketch / REGISTERS_PER_LONG + 1;
+  auto const results_iter = cudf::detail::make_counting_transform_iterator(0, [&](int i) {
+    return cudf::make_numeric_column(
+      cudf::data_type{cudf::type_id::INT64}, num_groups, cudf::mask_state::UNALLOCATED, stream, mr);
+  });
+  auto children =
+    std::vector<std::unique_ptr<cudf::column>>(results_iter, results_iter + num_long_cols);
+  auto d_results = [&] {
+    auto host_results_pointer_iter =
+      thrust::make_transform_iterator(children.begin(), [](auto const& results_column) {
+        return results_column->mutable_view().template data<int64_t>();
+      });
+    auto host_results_pointers =
+      std::vector<int64_t*>(host_results_pointer_iter, host_results_pointer_iter + children.size());
+    return cudf::detail::make_device_uvector_async(host_results_pointers, stream, mr);
+  }();
```

Comment:
Similarly, `host_results_pointers` can be destroyed before data is copied to device. Just remove this code block.
```suggestion
    auto host_results_pointer_iter =
      thrust::make_transform_iterator(children.begin(), [](auto const& results_column) {
        return results_column->mutable_view().template data<int64_t>();
      });
    auto host_results_pointers =
      std::vector<int64_t*>(host_results_pointer_iter, host_results_pointer_iter + children.size());
  auto d_results = cudf::detail::make_device_uvector_async(host_results_pointers, stream, mr);
```

Reply:
done